sponsorship_compassion/models/contracts.py

```diff
@@ -1110,98 +1098,65 @@ def _on_language_changed(self):
+        if message.state == "failure":
+            if answer["Code"] in [5000, ]:
+                logger.warning(message.answer)
             else:
-                messages[i].unlink()
-                errors += sponsorships[i]
-        if errors:
-            logger.error(
-                "Could not cancel contracts with following global_id(s):"
-                ", ".join(errors.mapped("global_id"))
-            )
+                error_message = answer["Message"]
+                logger.error(error_message)
+                raise UserError(_("GMC returned an error :") + "\n" + error_message)
 
-            raise UserError(
-                _("The current commitment at GMC side could not be cancelled.")
-            )
-
-        cancelled_sponsorships.write({"global_id": False})
-        return cancelled_sponsorships
+        self.write({"global_id": False, "state": "terminated"})
+        self.env.cr.commit()
+        return True
 
-    @api.multi
-    def _on_correspondant_changed(self):
-        """
-        This is useful for not having to internally cancel and create
-        a new commitment just to change the corresponding partner.
-        It will however cancel the commitment at GMC side and create a new
-        one.
-        But in Odoo, we will not see the commitment has changed.
-        """
+    def _add_correspondent(self, new_correspondents):
         message_obj = self.env["gmc.message"].with_context(async_mode=False)
         create_action = self.env.ref("sponsorship_compassion.create_sponsorship")
-
-        # Upsert correspondents
-        self.mapped("correspondent_id").upsert_constituent().process_messages()
+        self.correspondent_id.upsert_constituent().process_messages()
 
         # Create new sponsorships at GMC
-        messages = message_obj
-        for sponsorship in self:
-            partner = sponsorship.correspondent_id
-            messages += message_obj.create(
-                {
-                    "action_id": create_action.id,
-                    "child_id": sponsorship.child_id.id,
-                    "partner_id": partner.id,
-                    "object_id": sponsorship.id,
-                }
-            )
-        messages.process_messages()
-        for i in range(0, len(messages)):
-            if not messages[i].state == "success":
-                self[i].message_post(
-                    body=messages[i].failure_reason,
-                    subject=_("The sponsorship is no more active!")
-                )
+        message = message_obj.create(
+            {
+                "action_id": create_action.id,
+                "child_id": self.child_id.id,
+                "partner_id": new_correspondents.id,
+                "object_id": self.id,
+            }
+        )
+        message.process_messages()
+        if not (message.state == "success"):
+            self.message_post(body=message.failure_reason, subject=_("The sponsorship is no more active!"))
```

Comment:
This is not useful because the raise just after will delete this message (by rolling back any changes).